apps/dcos_dns/src/dcos_dns_mesos.erl

```diff
@@ -0,0 +1,384 @@
+    dns_records(DName, [AgentIP]).
+
+-spec(task_containerip(task()) -> [dns:dns_rr()]).
+task_containerip(#{name := Name, framework := Fwrk, task_ip := TaskIPs}) ->
+    DName = format_name([Name, Fwrk, <<"containerip">>], ?DCOS_DOMAIN),
+    dns_records(DName, TaskIPs);
+task_containerip(_Task) ->
+    [].
+
+-spec(task_autoip(task()) -> [dns:dns_rr()]).
+task_autoip(#{name := Name, framework := Fwrk,
+              agent_ip := AgentIP, task_ip := TaskIPs} = Task) ->
+    %% if task.port_mappings then agent_ip else task_ip
+    DName = format_name([Name, Fwrk, <<"autoip">>], ?DCOS_DOMAIN),
+    Ports = maps:get(ports, Task, []),
+    dns_records(DName,
+        case lists:any(fun is_port_mapping/1, Ports) of
+            true -> [AgentIP];
+            false -> TaskIPs
+        end
+    ).
+
+-spec(is_port_mapping(dcos_net_mesos_state:task_port()) -> boolean()).
+is_port_mapping(#{host_port := _HPort}) ->
+    true;
+is_port_mapping(_Port) ->
+    false.
+
+%%%===================================================================
+%%% Masters functions
+%%%===================================================================
+
+-spec(handle_masters(state()) -> state()).
+handle_masters(#state{masters=MRRs}=State) ->
+    ZoneName = ?DCOS_DOMAIN,
+    MRRs0 = master_records(ZoneName),
+
+    {NewRRs, OldRRs} = complement(MRRs0, MRRs),
+    lists:foreach(fun (#dns_rr{data=#dns_rrdata_a{ip = IP}}) ->
+        lager:notice("master ~p was added", [IP])
+    end, NewRRs),
+    lists:foreach(fun (#dns_rr{data=#dns_rrdata_a{ip = IP}}) ->
+        lager:notice("master ~p was removed", [IP])
+    end, OldRRs),
+
+    Ref = start_masters_timer(),
+    AddOps = [{add_all, NewRRs} || NewRRs =/= []],
+    RemoveOps = [{remove_all, OldRRs} || OldRRs =/= []],
+    State0 = State#state{masters_ref=Ref, masters=MRRs0},
+    handle_push_ops(AddOps ++ RemoveOps, State0).
+
+-spec(master_records(dns:dname()) -> [dns:dns_rr()]).
+master_records(ZoneName) ->
+    Masters = [IP || {IP, _} <- dcos_dns_config:mesos_resolvers()],
+    dns_records(<<"master.", ZoneName/binary>>, Masters).
+
+-spec(leader_records(dns:dname()) -> dns:dns_rr()).
+leader_records(ZoneName) ->
+    IP = dcos_net_dist:nodeip(),
+    dns_record(<<"leader.", ZoneName/binary>>, IP).
+
+-spec(start_masters_timer() -> reference()).
+start_masters_timer() ->
+    Timeout = application:get_env(dcos_dns, masters_timeout, 5000),
+    erlang:start_timer(Timeout, self(), masters).
+
+%%%===================================================================
+%%% DNS functions
+%%%===================================================================
+
+-spec(push_tasks(#{task_id() => task()}) ->
+    {ok, New :: [dns:dns_rr()], Old :: [dns:dns_rr()]}).
+push_tasks(Tasks) ->
+    ZoneName = ?DCOS_DOMAIN,
+    Records = maps:values(Tasks),
+    Records0 =
+        lists:flatten([
+            Records,
+            zone_records(ZoneName),
+            leader_records(ZoneName)
```

Comment:
is this a new record that we are adding because currently, we do not have a "leader" record ending with "dcos" zoneName

Reply:
These records are new. Initially I added them when `.mesos.thisdcos.direcory` zone was removed, but I hope at the moment it's fine to keep them as is.